syntax = "proto3";

package test;

import "google/protobuf/duration.proto";
import "stout/v1/flag.proto";
import "stout/v1/subcommand.proto";

///////////////////////////////////////////////////////////////////////////////

message Flags {
  string foo = 1 [
    (stout.v1.flag) = {
      names: [ "foo" ]
      help: "help"
      required: true
    }
  ];

  string _s = 2 [
    (stout.v1.flag) = {
      names: [ "s" ]
      help: "help"
      required: false
    }
  ];

  bool bar = 3 [
    (stout.v1.flag) = {
      names: [ "bar" ]
      help: "help"
    }
  ];

  int32 baz = 4 [
    (stout.v1.flag) = {
      names: [ "baz" ]
      help: "help"
    }
  ];

  google.protobuf.Duration duration = 5 [
    (stout.v1.flag) = {
      names: [ "duration" ]
      help: "help"
    }
  ];
}

///////////////////////////////////////////////////////////////////////////////

message MissingFlagName {
  string s = 1 [
    (stout.v1.flag) = {
      help: "help"
    }
  ];
}

///////////////////////////////////////////////////////////////////////////////

message MissingFlagHelp {
  string s = 1 [
    (stout.v1.flag) = {
      names: "name"
    }
  ];
}

///////////////////////////////////////////////////////////////////////////////

message DuplicateFlagName {
  string s1 = 1 [
    (stout.v1.flag) = {
      names: [ "same" ]
      help: "help"
    }
  ];

  string s2 = 2 [
    (stout.v1.flag) = {
      names: [ "other", "same" ]
      help: "help"
    }
  ];
}

///////////////////////////////////////////////////////////////////////////////

message DuplicateFlagsDeath {
  string s = 1 [
    (stout.v1.flag) = {
      names: [ "s" ]
      help: "help"
    }
  ];

  string ss = 2 [
    (stout.v1.flag) = {
      names: [ "ss", "ss_alias" ]
      help: "help"
    }
  ];

  bool duplicate = 3 [
    (stout.v1.flag) = {
      names: [ "duplicate", "other" ]
      help: "help"
    }
  ];

  bool b = 4 [
    (stout.v1.flag) = {
      names: [ "b" ]
      help: "help"
    }
  ];
}

///////////////////////////////////////////////////////////////////////////////

message BuildFlag {
  int32 other_flag = 1 [
    (stout.v1.flag) = {
      names: [ "other_flag" ]
      help: "help"
    }
  ];
}

///////////////////////////////////////////////////////////////////////////////

message InfoFlag {
  string info = 1 [
    (stout.v1.flag) = {
      names: [ "info" ]
      help: "help"
    }
  ];
}

///////////////////////////////////////////////////////////////////////////////

message SubcommandFlagsWithoutName {
  oneof subcommand {
    BuildFlag build = 1 [
      (stout.v1.subcommand) = {
        help: "help"
      }
    ];
    InfoFlag info_flag = 2 [
      (stout.v1.subcommand) = {
        names: [ "info_flag" ]
        help: "help"
      }
    ];
  }
}

///////////////////////////////////////////////////////////////////////////////

message SubcommandFlagsWithoutHelp {
  oneof subcommand {
    BuildFlag build = 1 [
      (stout.v1.subcommand) = {
        names: [ "build" ]
        help: "help"
      }
    ];
    InfoFlag info_flag = 2 [
      (stout.v1.subcommand) = {
        names: [ "info_flag" ]
      }
    ];
  }
}

///////////////////////////////////////////////////////////////////////////////

message SubcommandFlagsWithoutExtension {
  bool b = 1 [
    (stout.v1.flag) = {
      names: [ "b" ]
      help: "help"
    }
  ];

  oneof subcommand {
    BuildFlag build = 2 [
      (stout.v1.subcommand) = {
        names: [ "build" ]
        help: "help"
      }
    ];
    InfoFlag info_flag = 3;
  }
}

///////////////////////////////////////////////////////////////////////////////

message FlagsWithIncorrectExtension {
  bool b = 1 [
    (stout.v1.subcommand) = {
      names: [ "build" ]
      help: "help"
    }
  ];

  oneof subcommand {
    BuildFlag build = 2 [
      (stout.v1.subcommand) = {
        names: [ "build" ]
        help: "help"
      }
    ];
    InfoFlag info_flag = 3 [
      (stout.v1.subcommand) = {
        names: [ "info_flag" ]
        help: "help"
      }
    ];
  }
}

///////////////////////////////////////////////////////////////////////////////

message IncorrectOneofName {
  oneof other {
    BuildFlag build = 1 [
      (stout.v1.subcommand) = {
        names: [ "build" ]
        help: "help"
      }
    ];
    InfoFlag info_flag = 2 [
      (stout.v1.subcommand) = {
        names: [ "info_flag" ]
        help: "help"
      }
    ];
  }
}

///////////////////////////////////////////////////////////////////////////////

message SubcommandFlagExtension {
  oneof subcommand {
    BuildFlag build = 1 [
      (stout.v1.flag) = {
        names: [ "build" ]
        help: "help"
      }
    ];
    InfoFlag info_flag = 2 [
      (stout.v1.subcommand) = {
        names: [ "info_flag" ]
        help: "help"
      }
    ];
  }
}

///////////////////////////////////////////////////////////////////////////////
